### PR TITLE
Add authentication support via sshauthentication-api

### PR DESCRIPTION
**Pull request by hagau**
This implements support for external authentication via [sshauthentication-api](https://github.com/open-keychain/open-keychain/tree/master/sshauthentication-api) (e.g. [OpenKeychain](https://github.com/open-keychain/open-keychain))

This requires [sshlib #34](https://github.com/connectbot/sshlib/pull/34) and [sshlib #35](https://github.com/connectbot/sshlib/pull/35).

### CODE REVIEW

File: app/build.gradle [hunk 1]
```diff
@@ -36,6 +36,9 @@ repositories {
     maven {
         url 'https://oss.jfrog.org/artifactory/oss-snapshot-local/'
     }
+	maven {
+        url "https://dl.bintray.com/sufficientlysecure/maven"
```
Comment (dschuermann) [inline app/build.gradle]: I just requested the inclusion into JCenter. Then this is no longer necessary.

**Comment by dschuermann** [comment]
Related issues: https://github.com/connectbot/connectbot/issues/405 https://github.com/connectbot/connectbot/issues/510
in our project: https://github.com/open-keychain/open-keychain/issues/908

File: app/build.gradle (hunk 1 above)
Comment (hagau) [reply app/build.gradle]: Thanks, I will monitor the availability and remove the repo inclusion when available.

Comment (dschuermann) [reply app/build.gradle]: It's now on JCenter, my request has been approved.

Comment (hagau) [reply app/build.gradle]: Thanks, I have removed the superfluous maven repo.

**Comment by Valodim** [comment]
It's worth mentioning that all necessary changes have been merged and released in OpenKeychain since December, so no debug build is necessary for testing.

**Comment by Librazy** [comment]
I just can't wait it and built a apk out of openkeychain's fork (dev branches) and works perfectly. I can use openkeychain as ssh agent and login into my servers by that. Waiting for merge!

**Comment by jklein24** [comment]
Yes

**Comment by jklein24** [comment]
Yes

**Comment by ngharo** [comment]
I'm also anxious to see this merged. Is there any help you guys need in moving this forward?

**Comment by genofire** [comment]
When could i download it from F-Droid? ;)

**Comment by michib** [comment]
I build it for myself and was successfully able to connect to my server with usb-otg and yubikey stored ssh-key

**Comment by jplitza** [comment]
I was also able to compile and test this (with a few problems, see my line comment in connectbot/sshlib#34).
The workflow for selecting the external provider/key is a bit strange, however.

What I expected:
1. Go to "Edit host" dialog
2. Open "Use pubkey authentication" dropdown
3. Select "Use external agent"
4. Select "OpenKeychain" in the "Select Agent" dialog and maybe hit "Select"
5. Grant permission to use ConnectBot with OpenKeychain
6. Select which key to use

However, this is what happened:
1. Go to "Edit host" dialog
2. Open "Use pubkey authentication" dropdown
3. Select "Use external agent"
4. Grant permission to use ConnectBot with OpenKeychain
5. Select which key to use
6. Be left with the "Select Agent" dialog, with the "Select" button grayed our. In the background, however, everything looks alright already: The details of the selected key are visible below "Use pubkey authentication"
7. Hit "Cancel" in the "Select Agent" dialog.

**Comment by ngharo** [comment]
Is anyone interested in forking the project and moving it forward? There hasn't been any activity in a while and judging by the backlog of pull requests I'm not sure we'll ever see this merged here.

Disclaimer: I'm just a user.

**Comment by leonerd** [comment]
> Is anyone interested in forking the project and moving it forward? There hasn't been any activity in a while and judging by the backlog of pull requests I'm not sure we'll ever see this merged here.

> Disclaimer: I'm just a user.

Definitely. I would be and I'm one of the co-maintainers. I just don't know anything about Android development, nor do I have the upload/signing keys.

**Comment by hagau** [comment]
Thanks for testing everybody! :heart:

**Comment by genofire** [comment]
@hagau :-1:  F-Droid or other free repository?

**Comment by dschuermann** [comment]
@genofire Someone needs to create a metadata file for F-Droid to have it available there. We didn't also had time to do that. But you are free to do so.

**Comment by wiktor-k** [comment]
I'm glad that the fork happened as ConnectBot is clearly unmaintained but I see some issues with TermBot (key selection pops up twice, there is separate toggle for using external agent that is not used/saved, and my tests with Yubikey 4C did not work). Does TermBot have an issue tracker where this could be reported?

**Comment by dschuermann** [comment]
@wiktor-k TermBot requires OpenKeychain Version 5.0.1. The minor version is important. Also: which algorithm did you use with Yubikey 4C?

**Comment by wiktor-k** [comment]
@dschuermann Thanks for a quick reply! I'm using version 5.0.2. The key on Y4C is 4096 bit RSA. Good idea with the algorithms, I've got the following settings in `/etc/ssh/sshd_config`:
```
KexAlgorithms curve25519-sha256@libssh.org,diffie-hellman-group-exchange-sha256
Ciphers chacha20-poly1305@openssh.com,aes256-gcm@openssh.com,aes128-gcm@openssh.com
MACs hmac-sha2-512-etm@openssh.com,hmac-sha2-256-etm@openssh.com,umac-128-etm@openssh.com,hmac-sha2-512,hmac-sha2-256,umac-128@openssh.com
```
Is there a list of algorithms supported by TermBot/OpenKeychain?

**Comment by hagau** [comment]
@wiktor-k 
'HostKeyAlgorithms' would be the right config variable to look at. Try temporally commenting it out if set.

ECDSA, RSA, EdDSA & DSA are all supported with reasonable key sizes (no DSA-1024 keys).

**Comment by phryneas** [comment]
hagau: is this fork available on github? I could not find the repo.

**Comment by hagau** [comment]
@phryneas 
https://github.com/open-keychain/termbot

**Comment by hagau** [comment]
@jplitza @wiktor-k 
Thanks for testing!
There is now a issue tracker available at https://github.com/open-keychain/termbot
Please report any bugs there, especially the double key selection popup, since I can't reproduce that here.

**Comment by colemickens** [comment]
What's the state of this? Is there a desire to see this merged and finished, or is TermBot going to be around for the long haul at this point?

**Comment by dschuermann** [comment]
Our newest TermBot release is now using our Hardware Security SDK: https://hwsecurity.dev/
So OpenKeychain is no longer a requirement to get support for Security Keys and smart cards.

We think that this makes it extremely easy, even for novices, to use external security hardware: You no longer need to work with PGP key files. It only requires the hardware.

New page: https://hwsecurity.dev/termbot/
New source: https://github.com/cotechde/termbot

We plan to release the hwsecurity-openpgp artifact as GPLv3 in the near future. Key generation is also planned.

Easily add new hardare in TermBot itself, the SDK automatically detects device name and serial number:

**Comment by countsudoku** [comment]
Why was that closed? This was not merged and the TermBot approach is different to supporting the OpenKeychain API.

**Comment by mhalano** [comment]
This already was merged?

**Comment by genofire** [comment]
In 1.9.8 is it not.

**Comment by mhalano** [comment]
@hagau Any reason why did you close this PR? Are you working on a newer, improved version?

**Comment by maltalef101** [comment]
@hagau  why was this closed?

**Pull request closed without merging**